controller/resource.go

```diff
@@ -146,48 +179,209 @@ func (s ServiceStateLister) fetchECSTasks() (map[string]struct{}, error) {
+// exists and the cross-partition/cross-namespace read policy exists.
+func (s ServiceStateLister) ReconcileNamespaces(resources []Resource) error {
+	if !PartitionsEnabled(s.Partition) {
+		return nil
+	}
+
+	// create the cross-namespace read policy
+	if err := s.upsertCrossNSPolicy(); err != nil {
+		return err
+	}
+
+	// retrieve the list of unique namespaces from the map of services
+	ns := uniqueNamespaces(resources)
+	if len(ns) < 2 {
+		// only the default namespace exists, so nothing to do
+		return nil
+	}
+
+	if err := s.createNamespaces(ns); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// upsertCrossNSPolicy creates the cross-namespace read policy in the local
+// partition and default namespace, if it does not already exist.
+func (s ServiceStateLister) upsertCrossNSPolicy() error {
+	policy, _, err := s.ConsulClient.ACL().PolicyReadByName(
+		xnsPolicyName,
+		&api.QueryOptions{Partition: s.Partition, Namespace: DefaultNamespace},
+	)
+	if err != nil && !IsACLNotFoundError(err) {
+		return fmt.Errorf("reading cross-namespace policy: %w", err)
+	}
+
+	if policy != nil {
+		// the policy already exists so return.
+		return nil
+	}
+
+	// create the policy in the local partition and default namespace.
+	_, _, err = s.ConsulClient.ACL().PolicyCreate(&api.ACLPolicy{
+		Name:        xnsPolicyName,
+		Description: xnsPolicyDesc,
+		Partition:   s.Partition,
+		Namespace:   DefaultNamespace,
+		Rules:       fmt.Sprintf(xnsPolicyTpl, s.Partition),
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("creating cross-namespace policy: %w", err)
+	}
+
+	s.Log.Info("created cross-namespace policy", xnsPolicyName)
+	return nil
+}
+
+// createNamespaces creates each namespace in the list.
+// It also attaches the cross-namespace read policy to the namespace so that
+// each token created in the namespace inherits that policy.
+//
+// It does nothing for namespaces that already exist.
+func (s ServiceStateLister) createNamespaces(ns []string) error {
+	// retrieve the list of existing namespaces
+	existingNS, _, err := s.ConsulClient.Namespaces().List(&api.QueryOptions{Partition: s.Partition})
+	if err != nil {
+		return err
+	}
+
+	// create any namespaces that do not already exist
+	create := make([]string, 0, len(ns))
+	for _, n := range ns {
+		for _, e := range existingNS {
+			if e.Name != n {
+				create = append(create, n)
+				break
+			}
+		}
+	}
+	failed := make([]string, 0, len(create))
+	for _, n := range create {
+		s.Log.Info("creating namespace", n)
+		_, _, err = s.ConsulClient.Namespaces().Create(&api.Namespace{
+			Partition: s.Partition,
+			Name:      n,
+			ACLs:      &api.NamespaceACLConfig{PolicyDefaults: []api.ACLLink{{Name: xnsPolicyName}}},
+		}, nil)
+		if err != nil {
+			s.Log.Error("creating namespace", n, err)
+			failed = append(failed, n)
+		}
+	}
+	if len(failed) > 0 {
+		return fmt.Errorf("namespace creation failed for %s", strings.Join(failed, ","))
```

Comment:
nit: In other places, we've used [`go-multierror`](https://github.com/hashicorp/go-multierror) for collecting multiple errors.

https://github.com/hashicorp/consul-ecs/blob/0ad5fee74c0223f55238b1b80b6cb40f17ed4878/subcommand/health-sync/command.go#L157-L174

Reply:
Thanks for the pointer.. I've updated the code to use `go-multierror`